File: messaging/whatsapp/src/secret_filter.ts
// Runtime secret filter for outbound WhatsApp text (ADR-033 §4).
//
// Every row the outbound drain sends (conversation replies, progress
// messages, acknowledgements, task results, reminders, job replies, media
// captions) passes through `filterOutbound` first. The rules are the ones
// tools/check-secrets.sh applies to commits, read at runtime from the same
// sources:
//
//   A. `.env` values (>= 6 characters, minus benign values), plus the
//      operator's home directory;
//   B. the gitignored `.claude/secret-strings` denylist (whole word,
//      case-insensitive);
//   C. PII patterns: email addresses, WhatsApp JIDs, E.164 phone numbers,
//      home-directory paths.
//
// Plus credential-shaped tokens (private keys, API tokens) that should never
// leave the machine in any message.
//
// How a match is handled depends on the message:
//
//   - Credentials — `.env` values whose key names a credential (TOKEN,
//     SECRET, KEY, PASSWORD, …) and credential-shaped tokens — are redacted
//     in every message.
//   - A progress message (model narration sent while a turn runs) with any
//     match from A, B or C is withheld entirely: progress is optional, and
//     narration never needs an identifier.
//   - A message to a group is shared with its members: matches from A, B and
//     C are redacted.
//   - A final reply, note or result in the operator's own DM keeps the other
//     values: the operator asked for them (a stand-up summary names clients,
//     a lookup returns a phone number), and they are the operator's data.
//
// A redaction replaces the value with "[redacted]" and appends a note with
// the count. The drain logs every hit by kind and count, never the value.

import fs from "node:fs";
import path from "node:path";

export type Audience = "operator-dm" | "shared";

export interface SecretRules {
  /** Redacted everywhere. */
  credentials: string[];
  /** .env identifier values and the home directory (substring match). */
  identifiers: string[];
  /** Denylist literals (whole word, case-insensitive). */
  words: string[];
}

export interface FilterResult {
  text: string;
  /** One entry per match: its kind (never the value). */
  hits: string[];
  /** True when the whole message must not be sent. */
  withheld: boolean;
}

const BENIGN_VALUES = new Set(["info", "debug", "warn", "error", "trace", "true", "false", "claude"]);
const CREDENTIAL_KEY = /(TOKEN|SECRET|PASSWORD|PASSWD|CREDENTIAL|PRIVATE|COOKIE|(^|_)KEY($|_)|API_?KEY|AUTH)/i;
/** Keys whose values are configuration, not identifiers. */
const CONFIG_KEY = /(^|_)(TZ|LOG|LEVEL|MODEL|PORT|BIN|PATH|MODE|INTERVAL|TIMEOUT|DELAY|SECONDS|MS)($|_)/i;

/** Placeholders tools/check-secrets.sh also ignores. */
const PLACEHOLDERS = /example\.(com|org|net)|5511999999999|you@|@example|\/path\/to\//i;

const PII: Array<[RegExp, string]> = [
  [/[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/g, "pii-email"],
  [/\+?[0-9]{6,}(?::[0-9]+)?@(?:s\.whatsapp\.net|g\.us|c\.us|lid)/g, "pii-whatsapp-jid"],
  [/(?:\+55|\+1)[0-9]{9,}/g, "pii-phone-e164"],
  [/\/(?:Users|home)\/[a-z][a-z0-9_-]{2,}\//g, "pii-home-path"],
];

/** How a credential shape's match is redacted: the whole match, or group 2
 *  after a kept label in group 1 (`random`: only when group 2 looks random). */
type Part = "whole" | "value" | "random";

/** Credential-shaped tokens, most specific first. Mirrors `shapes()` in
 *  core/src/secret_filter.rs; both run core/testdata/credential_vectors.json. */
const CREDENTIAL_SHAPES: Array<[RegExp, string, Part]> = [
  [/-----BEGIN [A-Z0-9 ]*PRIVATE KEY-----[\s\S]*?(?:-----END [A-Z0-9 ]*PRIVATE KEY-----|$)/g, "credential-private-key", "whole"],
  [/\beyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}/g, "credential-jwt", "whole"],
  [/\bsk-(?:ant-)?[A-Za-z0-9_-]{20,}/g, "credential-api-key", "whole"],
  [/\b[sr]k_(?:live|test)_[A-Za-z0-9]{16,}/g, "credential-stripe", "whole"],
  [/\bAIza[0-9A-Za-z_-]{35}/g, "credential-google", "whole"],
  [/\bglpat-[A-Za-z0-9_-]{20,}/g, "credential-gitlab", "whole"],
  [/\b(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9]{30,}/g, "credential-github", "whole"],
  [/\bgithub_pat_[A-Za-z0-9_]{30,}/g, "credential-github", "whole"],
  [/\bxox[abeoprs]-[A-Za-z0-9-]{10,}/g, "credential-slack", "whole"],
  [/\b(?:AKIA|ASIA)[0-9A-Z]{16}\b/g, "credential-aws", "whole"],
  [/(aws[A-Za-z0-9_ -]{0,20}secret[A-Za-z0-9_ -]{0,20}["']?\s*[:=]\s*["']?)([A-Za-z0-9/+=]{40})/gi, "credential-aws", "value"],
  [/\bBearer\s+[A-Za-z0-9._~+/=-]{20,}/g, "credential-bearer", "whole"],
  [
    /((?:\b|_)(?:password|passwd|pwd|passphrase|secret|client[_-]?secret|token|access[_-]?token|refresh[_-]?token|auth[_-]?token|api[_-]?key|apikey|access[_-]?key|secret[_-]?key|private[_-]?key)\b["']?\s*[:=]\s*["']?)([^\s"'<>,;]{8,})/gi,
    "credential-labeled",
    "random",
  ],
];

/** A labeled value counts as a secret when it has at least 8 characters,
 *  mixes at least two character classes, and has a Shannon entropy of at
 *  least 3 bits per character. Mirrors `looks_random` in Rust. */
export function looksRandom(v: string): boolean {
  const chars = Array.from(v);
  if (chars.length < 8) return false;
  const classes = [/[a-z]/, /[A-Z]/, /[0-9]/, /[^A-Za-z0-9]/].filter((re) => re.test(v)).length;
  if (classes < 2) return false;
  const counts = new Map<string, number>();
  for (const c of chars) counts.set(c, (counts.get(c) ?? 0) + 1);
  let h = 0;
  for (const k of counts.values()) {
    const p = k / chars.length;
    h -= p * Math.log2(p);
  }
  return h >= 3.0;
}

function redactShape(text: string, re: RegExp, kind: string, part: Part, hits: string[]): string {
  return text.replace(re, (m: string, g1?: string, g2?: string) => {
    if (part === "whole") {
      hits.push(kind);
      return REDACTED;
    }
    if (part === "random" && !looksRandom(g2 ?? "")) return m;
    hits.push(kind);
    return `${g1 ?? ""}${REDACTED}`;
  });
}

/** Redact credentials only (the `.env` credential values and the shapes).
 *  Pure. */
export function redactCredentials(text: string, rules: Pick<SecretRules, "credentials">): { text: string; hits: string[] } {
  const hits: string[] = [];
  let out = text;
  for (const v of rules.credentials) out = replaceLiteral(out, v, "env-credential", hits);
  for (const [re, kind, part] of CREDENTIAL_SHAPES) out = redactShape(out, re, kind, part, hits);
  return { text: out, hits };
}

function unquote(v: string): string {
  const t = v.trim();
  if ((t.startsWith('"') && t.endsWith('"')) || (t.startsWith("'") && t.endsWith("'"))) {
    return t.slice(1, -1);
  }
  return t;
}

/** Rules from `.env` text, denylist text and the home directory. Pure. */
export function buildRules(envText: string, denylistText: string, home: string): SecretRules {
  const credentials: string[] = [];
  const identifiers: string[] = [];
  for (const raw of envText.split("\n")) {
    const line = raw.trim();
    if (!line || line.startsWith("#")) continue;
    const eq = line.indexOf("=");
    if (eq < 0) continue;
    const key = line.slice(0, eq).replace(/^export\s+/, "").trim();
    const value = unquote(line.slice(eq + 1));
    for (const part of value.split(",").map((p) => p.trim())) {
      if (part.length < 6 || BENIGN_VALUES.has(part.toLowerCase())) continue;
      if (CREDENTIAL_KEY.test(key)) credentials.push(part);
      else if (!CONFIG_KEY.test(key)) identifiers.push(part);
    }
  }
  if (home.length >= 6) identifiers.push(home);
  const words = denylistText
    .split("\n")
    .map((l) => l.trim())
    .filter((l) => l && !l.startsWith("#"));
  // Longest first, so a value that contains another is redacted whole.
  const byLen = (a: string, b: string) => b.length - a.length;
  return { credentials: credentials.sort(byLen), identifiers: identifiers.sort(byLen), words };
}

function escapeRe(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

const REDACTED = "[redacted]";

function replaceLiteral(text: string, value: string, kind: string, hits: string[]): string {
  if (!value || !text.includes(value)) return text;
  const parts = text.split(value);
  for (let i = 1; i < parts.length; i++) hits.push(kind);
  return parts.join(REDACTED);
}

function replaceRe(text: string, re: RegExp, kind: string, hits: string[], skipPlaceholders: boolean): string {
  return text.replace(re, (m) => {
    if (skipPlaceholders && PLACEHOLDERS.test(m)) return m;
    hits.push(kind);
    return REDACTED;
  });
}

/** Apply the rules to one outbound text. Pure. */
export function filterOutbound(
  text: string,
  rules: SecretRules,
  opts: { audience: Audience; progress: boolean; note?: string },
): FilterResult {
  // Credentials: every message.
  const cred = redactCredentials(text, rules);
  const hits: string[] = cred.hits;
  let out = cred.text;

  const strict = opts.progress || opts.audience === "shared";
  if (strict) {
    for (const v of rules.identifiers) out = replaceLiteral(out, v, "env-value", hits);
    for (const w of rules.words) {
      const re = new RegExp(`(?<![\\p{L}\\p{N}_])${escapeRe(w)}(?![\\p{L}\\p{N}_])`, "giu");
      out = replaceRe(out, re, "denylist", hits, false);
    }
    for (const [re, kind] of PII) out = replaceRe(out, re, kind, hits, true);
  }

  if (opts.progress && hits.length > 0) return { text: "", hits, withheld: true };
  if (hits.length > 0) {
    const note = opts.note ?? "({count} value(s) withheld by the secret filter)";
    out += `\n\n${note.replace("{count}", String(hits.length))}`;
  }
  return { text: out, hits, withheld: false };
}

/** Rules loaded from the workspace, re-read when a source file changes. */
export class SecretRuleSource {
  private rules: SecretRules = { credentials: [], identifiers: [], words: [] };
  private stamp = "";

  constructor(
    private readonly workspaceRoot: string,
    private readonly home: string = process.env.HOME ?? "",
  ) {}

  private files(): string[] {
    return [path.join(this.workspaceRoot, ".env"), path.join(this.workspaceRoot, ".claude/secret-strings")];
  }

  current(): SecretRules {
    const stamp = this.files()
      .map((f) => {
        try {
          const st = fs.statSync(f);
          return `${f}:${st.mtimeMs}:${st.size}`;
        } catch {
          return `${f}:-`;
        }
      })
      .join("|");
    if (stamp !== this.stamp) {
      const read = (f: string) => {
        try {
          return fs.readFileSync(f, "utf8");
        } catch {
          return "";
        }
      };
      const [env, deny] = this.files().map(read);
      this.rules = buildRules(env, deny, this.home);
      this.stamp = stamp;
    }
    return this.rules;
  }
}

File: core/src/secret_filter.rs
//! Credential redaction for text Nucleus sends from Rust (ADR-033).
//!
//! The WhatsApp bot filters every outbound row itself
//! (`messaging/whatsapp/src/secret_filter.ts`). Text that Rust sends
//! directly — a task result posted to Discord — passes through this module
//! first. It applies the credential part of the same rules: `.env` values
//! whose key names a credential, and credential-shaped tokens. Both
//! implementations run the vectors in `core/testdata/credential_vectors.json`.

use regex::Regex;
use std::path::Path;
use std::sync::OnceLock;

const REDACTED: &str = "[redacted]";

/// How a shape's match is redacted.
#[derive(Clone, Copy)]
enum Part {
    /// The whole match.
    Whole,
    /// Group 2; group 1 (a label such as `password=`) stays.
    Value,
    /// Group 2 when it looks random ([`looks_random`]); group 1 stays.
    RandomValue,
}

struct Shape {
    re: Regex,
    kind: &'static str,
    part: Part,
}

/// Credential-shaped tokens, most specific first. Mirrors
/// `CREDENTIAL_SHAPES` in secret_filter.ts.
fn shapes() -> &'static [Shape] {
    static SHAPES: OnceLock<Vec<Shape>> = OnceLock::new();
    SHAPES.get_or_init(|| {
        let s = |re: &str, kind, part| Shape { re: Regex::new(re).expect("credential shape"), kind, part };
        vec![
            s(
                r"(?s)-----BEGIN [A-Z0-9 ]*PRIVATE KEY-----.*?(?:-----END [A-Z0-9 ]*PRIVATE KEY-----|\z)",
                "credential-private-key",
                Part::Whole,
            ),
            s(r"\beyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}", "credential-jwt", Part::Whole),
            s(r"\bsk-(?:ant-)?[A-Za-z0-9_-]{20,}", "credential-api-key", Part::Whole),
            s(r"\b[sr]k_(?:live|test)_[A-Za-z0-9]{16,}", "credential-stripe", Part::Whole),
            s(r"\bAIza[0-9A-Za-z_-]{35}", "credential-google", Part::Whole),
            s(r"\bglpat-[A-Za-z0-9_-]{20,}", "credential-gitlab", Part::Whole),
            s(r"\b(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9]{30,}", "credential-github", Part::Whole),
            s(r"\bgithub_pat_[A-Za-z0-9_]{30,}", "credential-github", Part::Whole),
            s(r"\bxox[abeoprs]-[A-Za-z0-9-]{10,}", "credential-slack", Part::Whole),
            s(r"\b(?:AKIA|ASIA)[0-9A-Z]{16}\b", "credential-aws", Part::Whole),
            s(
                r#"(?i)(aws[A-Za-z0-9_ -]{0,20}secret[A-Za-z0-9_ -]{0,20}["']?\s*[:=]\s*["']?)([A-Za-z0-9/+=]{40})"#,
                "credential-aws",
                Part::Value,
            ),
            s(r"\bBearer\s+[A-Za-z0-9._~+/=-]{20,}", "credential-bearer", Part::Whole),
            s(
                r#"(?i)((?:\b|_)(?:password|passwd|pwd|passphrase|secret|client[_-]?secret|token|access[_-]?token|refresh[_-]?token|auth[_-]?token|api[_-]?key|apikey|access[_-]?key|secret[_-]?key|private[_-]?key)\b["']?\s*[:=]\s*["']?)([^\s"'<>,;]{8,})"#,
                "credential-labeled",
                Part::RandomValue,
            ),
        ]
    })
}

/// A labeled value counts as a secret when it is at least 8 characters,
/// mixes at least two character classes (lowercase, uppercase, digit,
/// other), and has a Shannon entropy of at least 3 bits per character.
/// Mirrors `looksRandom` in secret_filter.ts.
pub fn looks_random(v: &str) -> bool {
    let chars: Vec<char> = v.chars().collect();
    if chars.len() < 8 {
        return false;
    }
    let classes = [
        chars.iter().any(|c| c.is_ascii_lowercase()),
        chars.iter().any(|c| c.is_ascii_uppercase()),
        chars.iter().any(|c| c.is_ascii_digit()),
        chars.iter().any(|c| !c.is_ascii_alphanumeric()),
    ]
    .iter()
    .filter(|b| **b)
    .count();
    classes >= 2 && entropy(&chars) >= 3.0
}

fn entropy(chars: &[char]) -> f64 {
    let mut counts = std::collections::HashMap::new();
    for c in chars {
        *counts.entry(*c).or_insert(0usize) += 1;
    }
    let n = chars.len() as f64;
    counts.values().map(|&k| {
        let p = k as f64 / n;
        -p * p.log2()
    }).sum()
}

/// Result of [`CredentialRules::redact`].
pub struct Redacted {
    pub text: String,
    /// One entry per redaction: its kind, never the value.
    pub hits: Vec<&'static str>,
}

/// `.env` credential values plus the credential shapes.
#[derive(Default)]
pub struct CredentialRules {
    values: Vec<String>,
}

const BENIGN: &[&str] = &["info", "debug", "warn", "error", "trace", "true", "false", "claude"];

fn credential_key(key: &str) -> bool {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| {
        Regex::new(r"(?i)(TOKEN|SECRET|PASSWORD|PASSWD|CREDENTIAL|PRIVATE|COOKIE|(^|_)KEY($|_)|API_?KEY|AUTH)").unwrap()
    })
    .is_match(key)
}

impl CredentialRules {
    /// Credential values of `.env` text (keys that name a credential; values
    /// of at least 6 characters). Mirrors `buildRules` in secret_filter.ts.
    pub fn from_env_text(env: &str) -> Self {
        let mut values = Vec::new();
        for line in env.lines().map(str::trim) {
            if line.is_empty() || line.starts_with('#') {
                continue;
            }
            let Some((k, v)) = line.split_once('=') else { continue };
            let key = k.trim().trim_start_matches("export ").trim();
            if !credential_key(key) {
                continue;
            }
            let v = v.trim();
            let v = v
                .strip_prefix('"')
                .and_then(|x| x.strip_suffix('"'))
                .or_else(|| v.strip_prefix('\'').and_then(|x| x.strip_suffix('\'')))
                .unwrap_or(v);
            for part in v.split(',').map(str::trim) {
                if part.chars().count() >= 6 && !BENIGN.contains(&part.to_lowercase().as_str()) {
                    values.push(part.to_string());
                }
            }
        }
        values.sort_by_key(|v| std::cmp::Reverse(v.len()));
        CredentialRules { values }
    }

    /// Rules from `<workspace>/.env` (none when it is missing).
    pub fn from_workspace(workspace_root: &Path) -> Self {
        Self::from_env_text(&std::fs::read_to_string(workspace_root.join(".env")).unwrap_or_default())
    }

    /// Redact every credential in `text`.
    pub fn redact(&self, text: &str) -> Redacted {
        let mut hits = Vec::new();
        let mut out = text.to_string();
        for v in &self.values {
            let n = out.matches(v.as_str()).count();
            if n > 0 {
                out = out.replace(v.as_str(), REDACTED);
                hits.extend(std::iter::repeat("env-credential").take(n));
            }
        }
        for shape in shapes() {
            out = shape
                .re
                .replace_all(&out, |c: &regex::Captures| match shape.part {
                    Part::Whole => {
                        hits.push(shape.kind);
                        REDACTED.to_string()
                    }
                    Part::Value => {
                        hits.push(shape.kind);
                        format!("{}{REDACTED}", &c[1])
                    }
                    Part::RandomValue if looks_random(&c[2]) => {
                        hits.push(shape.kind);
                        format!("{}{REDACTED}", &c[1])
                    }
                    Part::RandomValue => c[0].to_string(),
                })
                .into_owned();
        }
        Redacted { text: out, hits }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Vector text is stored in pieces (`{"repeat": s, "times": n}` repeats
    /// `s`), so no committed file contains a contiguous credential-shaped
    /// string that secret scanners would flag.
    pub(crate) fn assemble(parts: &serde_json::Value) -> String {
        parts
            .as_array()
            .unwrap()
            .iter()
            .map(|p| match p {
                serde_json::Value::String(s) => s.clone(),
                o => o["repeat"].as_str().unwrap().repeat(o["times"].as_u64().unwrap() as usize),
            })
            .collect()
    }

    #[test]
    fn shared_vectors() {
        let raw = include_str!("../testdata/credential_vectors.json");
        let vectors: Vec<serde_json::Value> = serde_json::from_str(raw).unwrap();
        let rules = CredentialRules::default();
        assert!(vectors.len() >= 12);
        for v in vectors {
            let name = v["name"].as_str().unwrap();
            let text = assemble(&v["text"]);
            let r = rules.redact(&text);
            match v["kind"].as_str() {
                Some(kind) => {
                    assert!(r.hits.contains(&kind), "{name}: hits {:?} in {:?}", r.hits, r.text);
                    let secret = assemble(&v["secret"]);
                    assert!(!r.text.contains(&secret), "{name}: {:?}", r.text);
                }
                None => assert!(r.hits.is_empty(), "{name}: unexpected {:?} in {:?}", r.hits, r.text),
            }
            if let Some(keep) = v["keep"].as_str() {
                assert!(r.text.contains(keep), "{name}: {:?} lost {keep:?}", r.text);
            }
        }
    }

    #[test]
    fn env_credentials_are_redacted() {
        let rules = CredentialRules::from_env_text(
            "# c\nDISCORD_BOT_TOKEN=\"tok-abcdefghijklmnop\"\nNUCLEUS_TZ=Europe/Oslo\nAPI_KEY=short\n",
        );
        let r = rules.redact("a tok-abcdefghijklmnop b Europe/Oslo");
        assert_eq!(r.text, "a [redacted] b Europe/Oslo");
        assert_eq!(r.hits, vec!["env-credential"]);
    }
}
